### PR TITLE
SDL changes:

### PR DESCRIPTION
**Pull request by tmaul**
* add dsp to audio
* bodge in F1 for FFWD
* bodge in F12 for exit
* tweak opengl window sizing so it will try and make a best fit without being too big
* remove some unused stuff

### CODE REVIEW

**Comment by tmaul** [comment]
Don't merge this yet, I still have some more changes to make, and need to test on a few more targets. :)

**Comment by grant2258** [comment]
I trying to get the windows build going running into one issue though 

C:/repo/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/crt/crt0_c.c:18: undefined reference to `WinMain'
collect2.exe: error: ld returned 1 exit status
make[2]: *** [makefile.sdl:396: obj/cpu/m68k/m68kmake] Error 1
make[1]: *** [makefile.sdl:285: all] Error 2

since this is just to generate the 68k source im not sure what the best way to deal is.

**Comment by tmaul** [comment]
What is the output of:

 sdl-config --libs

**Comment by tmaul** [comment]
@grant2258 Probably in windows it needs the cflags of makefile.sdl changing to something like this:

CFLAGS = -pipe \
	   -std=gnu99 -O3 -ffloat-store \
	   -mwindows \
	   -fforce-addr -finline-limit=1200 -fthread-jumps \
	   -fexpensive-optimizations \
	   -Wall -Wno-long-long -Wno-sign-compare -Wno-uninitialized -Wno-unused \
	   -Wno-sequence-point -Wno-pedantic -m64 \
	   $(DEF) $(incdir)

**Comment by grant2258** [comment]
$ sdl-config  --libs
-L/mingw64/lib -lmingw32 -lSDLmain -lSDL -mwindows

$ sdl-config  --cflags
-I/mingw64/include/SDL -D_GNU_SOURCE=1 -Dmain=SDL_main

**Comment by grant2258** [comment]
fix for windows build 

https://github.com/grant2258/FBNeo/commit/de943a29487dd631203fd24071f6147ced0297b3

**Comment by tmaul** [comment]
@grant2258 thanks, I have merged in the changes

**Comment by grant2258** [comment]
np at all just compiled your branch there is an issue with ini file not reading rom paths not longer work and no ini is created on the make sdl

**Comment by grant2258** [comment]
SDL2 compiles fine its running a bit fast and there is no audio on windows and the keyboard input needs some rework. Ill hook the joysticks up and look at the keyboard over the weekend unless you want to do it yourself

**Comment by tmaul** [comment]
@grant2258 nice! I suspected it might be related to the SDL to fbn key mapping, just had no time for coding.

The speed is also off in Linux so I will take a look at that. Also planning on redoing the video so it's just pure sdl2 rather than just a patched opengl sdl1.2 renderer. 

If you do look at the joystick stuff, the SDL 2 APIs look much nicer than the 1.2 ones, so that's something :)

**Comment by grant2258** [comment]
I managed to get the audio working is a windows sdl issue will send a patch after dinner to get it working. For now ill update the joysticks code we have if thats ok so the sdl1 port works ok as well. Then move onto a sdl2 implementation.

**Comment by grant2258** [comment]
patches fix the jumping fps and windows sound. https://github.com/grant2258/FBNeo/commit/5e8dca0c52b56473928a6eaf2f9c055d80cad244

**Comment by grant2258** [comment]
Changes are looking really good I spent a bit of today trying to get the gfx to work. Funny thing is it works in linux but in windows the height id well off. Looking at the code form earlier there is one thing I can point out one if the if conditions needs fixed up. https://github.com/grant2258/FBNeo/blob/63c7495eadcee208d7834f2b0a63aa995a4a55eb/src/intf/video/sdl/vid_sdl2.cpp#L31 its easy enough to fix just point it out from when i was testing.

and small updates for  vertical rotation games.

https://github.com/grant2258/FBNeo/commit/63c7495eadcee208d7834f2b0a63aa995a4a55eb

**Comment by grant2258** [comment]
When you have time can you look in your home/.config directory im getting a load of bad invalid encoding 2 folders made when i run fbneo.

**Comment by tmaul** [comment]
That's odd, I will take a look. Maybe need to find another cross platform
config thing. SDL2 looks like it has something that might work but the
documentation is a bit vague.

**Comment by grant2258** [comment]
CreateSavePath(cfgdir); is causing it have a few people round will have  a look whats going on at some point.

**Comment by tmaul** [comment]
Thanks, I have merged it all in. Just struggled to find time to do it  And don't worry at all about generating too many pull requests, that's what git is for :)

**Comment by tmaul** [comment]
Here's a question: 

For loading PNG files for previews, etc, Should I just use SDL_Image? I was thinking it might be worth it but then it's another dependency that some platforms might not have. Its either that or use a wrapper around libpng to load to an SDL_texture (or maybe even use https://github.com/nothings/stb/blob/master/stb_image.h then we can load other formats as well)

**Comment by grant2258** [comment]
either way would would work to be honest but sdl image would save a lot of hassle. If another system doesnt support sdlimage the other code could be done if/when needed.

**Comment by tmaul** [comment]
Sorry for all those commits, I seem to have a tab issue. Or I have fixed the tab issue I had before. Something is different since reinstalling anyway. Its either Atom or Ubuntu being annoying :)

**Comment by grant2258** [comment]
some of that code was the legacy code that I didnt remove inase you needed it. Vertical games no longer work on windows now. There is something off going on with the windows port. Ill sort this out to work on linux and do a manual rotation for windows.

**Comment by tmaul** [comment]
I'm happy if the code goes back in, if it fixes the problem. :)

**Comment by grant2258** [comment]
naw it was  band aid anyway best fixing it properly :)

**Comment by tmaul** [comment]
Might be a driver issue? I guess if it needs something that SDL2 is doing for rotation on textures. Could also be something not done in the right order or a call missing. SDL2 docs are a bit vague sometimes so I could have missed something.

**Comment by 0xe1f** [comment]
@tmaul 
`make pi` doesn't link:
Linking executable... fbneo
/usr/bin/ld: obj/burner/sdl/neocdlist.o: in function `SetNeoCDTitle(char*)':
neocdlist.cpp:(.text+0xd4): undefined reference to `szAppBurnVer'
/usr/bin/ld: obj/burner/cong.o: in function `ConfigGameSave(bool)':
cong.cpp:(.text+0x32c): undefined reference to `szAppBurnVer'
/usr/bin/ld: obj/burner/dat.o: in function `write_datfile(int, _IO_FILE*)':
dat.cpp:(.text+0x1150): undefined reference to `szAppBurnVer'
/usr/bin/ld: obj/burner/sshot.o: in function `MakeScreenShot()':
sshot.cpp:(.text+0x720): undefined reference to `szAppBurnVer'
/usr/bin/ld: obj/burner/sdl/config.o: in function `CreateConfigName(char*)':
config.cpp:(.text+0x24): undefined reference to `AppConfigPath(char*)'
/usr/bin/ld: obj/burner/pi/run_pi.o: in function `RunGetNextSound(int)':
run_pi.cpp:(.text+0x1f4): undefined reference to `bAppDoFast'
/usr/bin/ld: obj/burner/pi/run_pi.o: in function `RunIdle() [clone .part.2]':
run_pi.cpp:(.text+0x33c): undefined reference to `bAppDoFast'
collect2: error: ld returned 1 exit status
make[2]: *** [makefile.pi:335: fbneo] Error 1
make[1]: *** [makefile.pi:306: all] Error 2
make: *** [makefile:98: pi] Error 2

`make sdl2` no video, with vidselect 0 or 1
mac build: no issues

**Comment by tmaul** [comment]
@0xe1f thanks for checking, I will put a patch in for pi tonight. My pi3 is currently in use so I have not been able to test. We should set up cross complication for that target for CI at some point to stop me breaking things :)

**Comment by grant2258** [comment]
sdl2 works fine on the pi3 for me. havent looked at sdl1 or pi for it though

**Comment by tmaul** [comment]
I was wondering if we should make vsync enabled the default? Not sure one makes the most sense

**Comment by 0xe1f** [comment]
For the SDL build? As log as it can be disabled via .ini file or command line, vsync by default is probably a good idea.

**Comment by tmaul** [comment]
Yes, I was thinking of changing the -vsync option to -novsync. Also, should the command line options override the ini options but only for that session? Or is saving them to the ini file like it does currently a good idea?

**Comment by 0xe1f** [comment]
Just for the session. Saving the argument to ini file kills the perk of being able to temporarily override it via command line.

**Comment by grant2258** [comment]
am having some issues compiling the windows port ive fixed the winmain.h issues now there is just one left.

makefile.sdl2:491: obj/burner/un7z.d: No such file or directory
make[2]: *** No rule to make target 'obj/burner/un7z.d'.  Stop.
make[1]: *** [makefile.sdl2:304: all] Error 2
make: *** [makefile:92: sdl2] Error 2

**Comment by tmaul** [comment]
@grant2258 is that with or without the SKIPDEPEND=1 override?

**Comment by grant2258** [comment]
just issued a standard make sdl2 -j8 its probably something windows end to be hnest i just updated my msys

**Comment by tmaul** [comment]
I have seen that error when using appveyor, in the end I had to update the yml to pass SKIPDEPEND=1, so that might be worth a try

**Comment by grant2258** [comment]
cheers for that will give it a go

**Comment by grant2258** [comment]
E:/mingwbuild/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/crt/crt0_c.c:18: undefined reference to `WinMain'
collect2.exe: error: ld returned 1 exit status
make[1]: *** [makefile.sdl2:478: src/dep/generated/pgm_sprite.h] Error 1
make: *** [makefile:92: sdl2] Error 2

with SKIPDEPEND=1

**Comment by tmaul** [comment]
@grant2258 in makefile.sdl2 can you try editing the relevent bit to something like 

```
$(pgm_sprite.h):	pgm_sprite_create.cpp
	@echo Generating $(srcdir)dep/generated/$(@F)...
	@$(CC) $(CXXFLAGS)  -Dmain=main  $(LDFLAGS) $< \
		-o $(subst $(srcdir),$(objdir),$(<D))/$(<F:.cpp=.exe)
	@$(subst $(srcdir),$(objdir),$(<D))/$(<F:.cpp=.exe) >$@
```

**Comment by grant2258** [comment]
I Think i know what this give me a few to mess about with it going to grab your updates first before messing with the makefile.

**Comment by grant2258** [comment]
I got it compiling again but the sound is gone again from the windows port for some reason been looking at that cant see anything obvious will keep on looking before i put a pull req in is the sound working ok in linux?

**Comment by tmaul** [comment]
Yeah, it's working fine in all the games I've tried. Might be worth
disabling the DSP stuff just in case it's that?

**Comment by grant2258** [comment]
What i will say is the graphic look a lot less pixelated now unless its my imagination. I do have a small issue the vsync and fullscreen options dont seem to be working and can no longer resize the window.

**Comment by tmaul** [comment]
Hmm, that could be the filtering?

**Comment by grant2258** [comment]
Made a wee mistake after i done my fixes for windows to test the pull request. I removed the obj dir and done a make sdl instead of make sdl2 the command switches are working fine sorry about that.

**Comment by dinkc64** [comment]
nice work guys :)

**Comment by grant2258** [comment]
yea @tmaul has put all the hard grapht in had done a great job so far. Im just hacking in some joystick stuff for now to do some testing before getting to input. There are two particular bugs im looking at bombjack sound i think its just distortion from being too loud I could be wrong on this though. And gng the grass seems to flash yellow when scrolling. Ill need to test this all on linux as well to make sure its not a windows bug

**Comment by tmaul** [comment]
@dinkc64  Thanks Buddy :) If you consider the SDL port was started at least 12 years ago, it's about time it was polished up a bit lol

@grant2258 I wonder if that is related to it using the wrong bitdepth for the game? It might be worth trying to force the line nVidImageDepth = bDrvOkay ? 16 : 32; to always be 32 or 16 to see if that helps. I think it might actually need support for 15bpp for some drivers as well.

**Comment by tmaul** [comment]
@grant2258  It looks like that might be to do with the scaling code. Does it still do it if you pass -best when running? I think the ratio correction code is still wrong

**Comment by grant2258** [comment]
I noticed the softfx in stanalone is doing this yellow you really need pay attention because it happens a lot faster. 

If you start the game and scroll to the first hill with the ladder pay particular attention to the headstones on the top row while your walk by on the bottom sometimes the headstones appear to move across a little like they are offset wrong and move back  i see this on all ports it might be an effect of original hardware. 

I just paid attention to where i dont normally look with the yellow tint appearing. I might just be over looking into this one. I tried setting to 32 bit the result was the same. Dont get me wrong the game is still very much playable. Ive been playing it a lot lately. Since ive had video issues in windows before im going to compile this on the pi and check it there its the only linux box i have access too tonight.

File: src/burner/sdl/sdl2_gui.cpp [hunk 1]
```diff
@@ -294,12 +284,20 @@ void RefreshRomList(bool force_rescan)
 		}
 		BzipClose();
 
-		if (i%200==1)
+		if (i%200==1 || i == 0)
 		{
```
Comment (dinkc64) [inline src/burner/sdl/sdl2_gui.cpp]: "if (i % 200 == 0)" ? :)

**Comment by tmaul** [comment]
Indeed lol. And I was sober.

**Comment by tmaul** [comment]
FYI, @0xe1f When this eventually gets merged in (which hopefully won't be too many more days) then you will need to add cd_sdl.cpp to the xcode project. I am 90% sure it will work as is on the macos version once you have set up some way of selecting a CD :)

File: src/burner/sdl/main.cpp [hunk 2]
```diff
@@ -275,6 +275,7 @@ int main(int argc, char *argv[])
       printf("e.g.: %s -menu -joy\n", argv[0]);
       printf("For NeoCD games:\n");
       printf("%s neocdz -cd path/to/ccd/filename.ccd\n", argv[0]);
+      printf("Usage is resitricted by the licence at https://raw.githubusercontent.com/finalburnneo/FBNeo/master/src/license.txt\n");
```
Comment (dinkc64) [inline src/burner/sdl/main.cpp]: kev, some bad news re: ccd - a lot of the subchannel data for ccd is bad and causes skips in audio etc.  It's probably best to prefer/recommend using bin/cue (although still support .ccd behind the scenes)
Unfortunately I found this out _after_ putting so much time into supporting it properly, as well as researched the issue, did tests in other emus and talked with other devs about it.  luckily bin/cue is 100% solid, though :)

**Comment by grant2258** [comment]
There is one issue I am having with the menu at the moment in window the position is highlighted clear with the highlight scrolling through colors. Since im stating testing on the linux build on my barcade im noticing this hightlight isint working on retropie and x64 (linux mint) its very hard to see visually where you are on the list

**Comment by tmaul** [comment]
@grant2258 Could I see a screenshot please? 

Maybe we should change the background for the selected game as well, that might help a bit. Or maybe increase the size of the font via a command line option?

**Comment by grant2258** [comment]
I dont think its the background I recorded a video of what its doing in windows. I dont get this in linux though the color cycling isint happening.

https://drive.google.com/open?id=181R7qaQnRSkGcJEIZGXRhMdQPgXFfmQu 

The one thing i will note is the raspberry pi and linux box are on 1080p monitors and this laptop is 720p.

**Comment by tmaul** [comment]
Yeah that colour cycling is the same as it is here. If it's not happening on the laptop, that's something that needs to be looked at. Does it work in windowed mode? Also, if you run it without -fullscreen after editing the nVidGuiWidth and nVidGuiHeight defines to  something else does it still not cycle? and I was suggesting with the background something like putting a red background (with alpha channel) behind that line so it is still clear what the selected game is

**Comment by grant2258** [comment]
ok here is what im getting from the linux box im testing on. Ill need to get roms on this laptop to see with the list loaded on windows. The Strange thing is its working on the recording and not on the screen when im looking at it. 

https://drive.google.com/open?id=1czEKVLz4KBLOcOtlNX05qTfV3UXC3oA2

**Comment by tmaul** [comment]
@grant2258  Ah cool, I have added a bar behind the selected game to help if anyone else runs into this problem

**Comment by grant2258** [comment]
nice!

**Comment by tmaul** [comment]
We might need an SDL2 audio expert to find out what is up with the audio output. If I use the pulse audio output intf then I get (almost) 60fps in kinst, where as if I use aud_sdl.cpp then i get about 14fps, so something is wrong with it somewhere. The next time I have some free time I might try again from scratch, using the pulse audio intf as a base.

**Comment by grant2258** [comment]
Im missing this one dont have chds is there a particular version we need dont see any chd`s in the dat

**Comment by tmaul** [comment]
static struct BurnHDDInfo kinstHDDDesc[] = {
	{ "kinst.img",		0x7d01200, 0x2b9b6c0d }
};

It needs to be unzipped and placed in the HDD/kinst directory near where it's saving states, etc.

**Comment by Loggan08** [comment]
Just as a piece of info, the IMGs can be grabbed here : http://213.185.13.233/roms/FinalBurn%20Neo/support.7z

**Comment by grant2258** [comment]
grabbing now thanks for that been spending a bit of time looking into this v60 issue

**Comment by grant2258** [comment]
@tmaul ive been looking over the sdl_audio code and think ive found one issue.

https://github.com/finalburnneo/FBNeo/blob/dfd44893c21cc12d7d5d0f51077ac6e0ce3567bc/src/intf/audio/sdl/aud_sdl.cpp#L95-L114

This sets a local variable as a flag basically does nothing I think this should be calling RunFrame()

**Comment by tmaul** [comment]
Yeah, I thought that might have been causing the issue as well but it seemed to be working ok.

**Comment by grant2258** [comment]
good new well some at least

https://github.com/finalburnneo/FBNeo/blob/dfd44893c21cc12d7d5d0f51077ac6e0ce3567bc/src/intf/audio/sdl/aud_sdl.cpp#L102-L105

commenting this out fixes bombjack following the loop i think this code should be in GetNextSound

**Comment by tmaul** [comment]
@grant2258  What is the issue with bombjack? It seems to be ok here

**Comment by grant2258** [comment]
when you start the game the sound muffles i can record it before and after if you like its easier to hear it that way

**Comment by tmaul** [comment]
@grant2258 thanks, I'll take a look.

**Comment by grant2258** [comment]
no problems moving instead of commenting it out 

```		if (nAudDSPModule[0])
		{
			DspDo(nAudNextSound, nAudSegLen);
		}
```
before GetNextSound(bDraw); also fixes the problem

**Comment by tmaul** [comment]
For some reason the filter is causing clipping with bombjack. Might indicate some sort of buffer overflow but it could also just be too loud in the driver.  I'm going to leave the code there but disable the DSP module by default until I have had chance to look at the causes in more detail. You will need to edit your fbneo.ini to disable it in the mean time.

The other issue related to sound (apparently) is that the SDL audio intf is causing the games to render too many frames. It's something like 3 frames for every emulated frame. If I copy the pulse audio driver over it seems to work a lot better (about 10+ FPS minimum) which does indicate something is wrong somewhere in code we are using for SDL audio output.

**Comment by grant2258** [comment]
i know the sdl_delay will be contributing to it tit can range from 1 to 15 ticks put a pull req in to take that delay off

**Comment by tmaul** [comment]
@grant2258  Can you grab the latest code and try md_sonic3 - To me, it sounds totally wrong

**Comment by grant2258** [comment]
I sounds far too fast at saying that something must be locking us to 60fps(im guessing it could be the sound) or we are not setting something right sonic3 is pal and has this issue sonic3u is fine

**Comment by tmaul** [comment]
It's possible. Also, although the speed wasn't really the issue I was hearing. However, it looks like having vsync on was causing all the problems i was hearing. Turning it off seems to make everything run a lot better. I have no idea what the refresh rate of this monitor is but it's not 60hz it would seem :)

**Comment by tmaul** [comment]
Actually, with your last change and vsync off I am getting a pretty much locked in framerate accoring to the FPS counter

**Comment by grant2258** [comment]
yea your right sorry i had to go to my alt dat branch it wasnt up to date dont a rebase is running 60fps pal seems to be running too fast though is killer insitinc any better?

**Comment by tmaul** [comment]
It looks like the sound code is inited before it gets the required FPS from the game. I will see if I can fix that quickly.

**Comment by tmaul** [comment]
@grant2258 Check out the latest changes, seems to work a lot better

**Comment by grant2258** [comment]
yeap running 50 fps like we expect it to sonic that is. Looks like the sound is controlling the timing!

**Comment by tmaul** [comment]
@grant2258  Yes, the sound does control the timing, which is why it didn't make sense when it was running at 100's fps before. Still makes me wonder if something dodgy is going on with the audio out code.

**Comment by grant2258** [comment]
I thin the 2 sdl_delays threw things off as well fps not being set right.  its fine for me since i rebased my alt dat branch. ill keep an eye on the fps counter maybe you could leave it on by default for now so we can see any odd demons it honestly all looks good now to me. I fond startwars vector game seems to be running well

**Comment by grant2258** [comment]
@tmaul there is a bit of inconsistency with the paths in the sdl2 files when it comes to the png screenshot directory. its set here. 

https://github.com/finalburnneo/FBNeo/blob/aa97b456615c0ca44a3a87956160b558cb324bf0/src/burner/sdl/sdl2_gui.cpp#L49-L53

the get pref path returns ~/.local/share/  and C:\Users\username\AppData\Roaming\fbneo\

should these be using ther sdlgetprefpath as well or is this intentional ?

**Comment by tmaul** [comment]
@grant2258  This was intentional, as I thought it might be better if using the title screenshot pack by Gab75 (https://neo-source.com/index.php?topic=2838.0) if it were treated more like the roms just in case they were being used by something other than FBNeo

**Comment by grant2258** [comment]
ah no problems just asking was going to patch it if it wasnt the case. Thanks for the clarification

**Comment by tmaul** [comment]
@grant2258 It's always worth checking, at some point soon I would like to get this merged in with master so always point out anything that looks suspect :)

**Comment by grant2258** [comment]
Ill double check at the weekend its using the same support paths as burner.h thats the only check needed now i know it should be set to this variable and it will allow people to change it if they choose to in the ini file instead of being hard coded im on the right path now you pointed it out cheers!

**Comment by tmaul** [comment]
@grant2258 That doesn't sound too bad an idea. The ingame menu also needs some input (and dipswitch) settings. i was thinking of just having so it could map a controller to a player but I guess it also needs to remap buttons. I'll hold off looking at that until you have taken a look to see what you suggest is possible :)

**Comment by grant2258** [comment]
@tmaul looks like this is doable with the inputs we can get the joystick button bind for sdl1 need to look into it a little more though. https://wiki.libsdl.org/SDL_GameControllerButtonBind

**Comment by tmaul** [comment]
@grant2258 That looks like it would work. Also you were right about the -1 causing a missing line, I have committed that.

**Comment by grant2258** [comment]
yea i worked on it a little last night I just want to go with the flow with input controls had a bad experience before so dont want to step on any toes.

```
//joystick testing
#include "SDL.h"


int main(int argc, char *argv[])
{
   SDL_GameController *JoyList = NULL; // uses the same index as SDL_Joystick
   SDL_JoystickGUID GUID; //
   if (SDL_Init( SDL_INIT_VIDEO | SDL_INIT_JOYSTICK | SDL_INIT_GAMECONTROLLER) < 0)
   {
      fprintf(stderr, "Couldn't initialize SDL: %s\n", SDL_GetError());
      exit(1);
   }
    
   for (int i = 0; i < SDL_NumJoysticks(); ++i) 
   {
      const char *name = SDL_GameControllerNameForIndex(i);
      if (name) {
        printf("game controller index Joystick %i has game controller name '%s'\n", i, name);
      } 
      else 
      {
         printf("game controller Joystick %i has no game controller name.\n", i);
      }
      JoyList = SDL_JoystickOpen(i);
      printf("(legacy) Joystick %i name %s\n", i, SDL_JoystickName(JoyList) );

      char *mapping;
      JoyList = SDL_GameControllerOpen(i);


      mapping = SDL_GameControllerMapping(JoyList);
      SDL_GameControllerAddMapping(mapping);
      SDL_GameControllerButtonBind bind;

      SDL_Log("Controller %i is mapped as \"%s\".", i, mapping);

      bind = SDL_GameControllerGetBindForButton(JoyList, SDL_CONTROLLER_BUTTON_A );
      SDL_Log(" a = %d \n" , bind.value.button );

      bind = SDL_GameControllerGetBindForButton(JoyList, SDL_CONTROLLER_BUTTON_B);
      SDL_Log(" b = %d \n" , bind.value.button );

      bind = SDL_GameControllerGetBindForButton(JoyList, SDL_CONTROLLER_BUTTON_X );
      SDL_Log(" x = %d \n" , bind.value.button );

      bind = SDL_GameControllerGetBindForButton(JoyList, SDL_CONTROLLER_BUTTON_Y);
      SDL_Log(" y = %d \n" , bind.value.button );

      bind = SDL_GameControllerGetBindForButton(JoyList, SDL_CONTROLLER_BUTTON_LEFTSHOULDER  );
      SDL_Log(" l1 = %d \n" , bind.value.button );

      bind = SDL_GameControllerGetBindForButton(JoyList, SDL_CONTROLLER_BUTTON_RIGHTSHOULDER );
      SDL_Log(" r1 = %d \n" , bind.value.button );

      bind = SDL_GameControllerGetBindForButton(JoyList, SDL_CONTROLLER_BUTTON_BACK   );
      SDL_Log(" back = %d \n" , bind.value.button );

      bind = SDL_GameControllerGetBindForButton(JoyList, SDL_CONTROLLER_BUTTON_START  );
      SDL_Log(" start = %d \n" , bind.value.button );

      SDL_free(mapping);
   }
}
```
gets us the binds needs error checking all i need to do is integrate this into sdl2

**Comment by grant2258** [comment]
@dinkc64 have a quick question the spectrum maps the keyboard which really does need done for a computer. There is a problem with the msx though it doesnt seem to be setting the keyboard a few games need f1 to f4 to start. Is there any reason the keyboard isint mapped for this system?

**Comment by grant2258** [comment]
Need some input from people for default settings on input mappings for games like jackal, and midnight resistance, ikari warriors. For me i use the l/r triggers just want some input of what you all think and what settings you guys use for this.

**Comment by dinkc64** [comment]
At the time I wrote the msx driver, it wasn't possible to do shift chars -- for instance:
shift-; is :, and there were a few other issues, like getting the shift state.  The key callback in d_msx.cpp needs to be set to be called for every keypress on the keyboard.  You can see how I do it in burner/win32/run.cpp (search:  keycallback)
Also libretro has this hooked up too, be sure to check it out as well.

best regards,
- dink

**Comment by dinkc64** [comment]
@grant2258, for Jackal, midres, hbarrel, ikari (all rotate games) there is an alternate rotation option other than using l/r buttons for rotation.  Hold the "Rotate Button" and use the joystick/dpad to move your guy in the direction wanted.  This is how I play them.
Also midres has a special "HOLD" function for rotation while crawling or running backwards that is really handy :)  works like this:
hold the rotate button (#3) while crawling or running
this button will latch the direction you are moving until you let up on it, so you can keep moving while rotating your gun and fireing.

**Comment by grant2258** [comment]
actually the hold button sounds ideal i can go with that one sounds like a plan. I see the libretro port assigns fx buttons to the joypad.  These F(x) can be mapped i was meaning the rest of the keyboard. I can understand why the f(1) buttons would need special attention. I just want to set some sane default options for FBA without too many rules since arcades use the jamma standard. I got pretty distracted playing dizzy on teh spectrum yesterday! still plays well.

**Comment by tmaul** [comment]
@grant2258 I am trying to get it running on windows so I can change the directories, etc that it uses to bring it more in line with the native windows version. Does the current code work for you if you do a clean build? GDB doesn't seem to want to give me a good stack trace so i have no idea where it's crashing currently.

**Comment by grant2258** [comment]
ill have a look was fine last night

**Comment by tmaul** [comment]
@grant2258  I have figured out the problem, just doing some tests and will upload a new version that sorts out the windows paths a bit, as windows is much more unrestricted when it comes to file access, plus it keeps it in line with the windows native build

**Comment by grant2258** [comment]
Yea to be honest i think it would be better for the windows port to do it the same way.

**Comment by tmaul** [comment]
@grant2258  I have made a first pass on the windows port to try and bring it a bit more inline with the windows native version. Still needs some work. I need to sleep now, but I like your idea for linux. If you don't raise a pull request for that I will get to it at some point in the near future :)

**Comment by grant2258** [comment]
will get it at the weekend usual the only time i have to put at least and hour or two in.

**Comment by dinkc64** [comment]
@tmaul, 
DrMingw is a just-in-time debugger that will backcrawl the stack and help you find where it crashed.  Just make sure optimization is off aka -O0 and -ggdb3 flags on the compiler(s).
It helps me find most crashes, and saves so much time.  get it here:
https://github.com/jrfonseca/drmingw

tip, it's installer is weird, where you unzip it to is the install-point.  so if you want it in c:\mingw\drmingw, unzip it there, then go to c:\mingw\drmingw\bin and run drmingw -i
now test it:
Load up a gcc compiled fbn debug version windows ui.  press alt-c to involk a crash (hehe)
When FB crashes, click the [debug] button instead of [ok] 
DrMingw will spawn after a second (or 5) and show you the sourcecode and line exactly where it crashed.  it will get the code/line(s) wrong sometimes if you use -O[1..3], so again make sure -O0 (O Zero)
If everything is OK, you'll see:

fbaltc

**Comment by tmaul** [comment]
Thanks @dinkc64 that's very helpful :) it is much better than the mingw gdb I tried. Will help track down the last few crashes I saw.

**Comment by tmaul** [comment]
Screenshot from 2020-01-14 22-54-06
Can't get the scaling right but might be worth pursing...

**Comment by dinkc64** [comment]
tmaul glad DrMingw helped :)
and..
whaaaaaaaaaaaat?  you have a batsugun cabinet? :)

**Comment by tmaul** [comment]
Well, I have one I so much as I downloaded an overlay of one :) tbh it's the only overlay that looks half good currently lol

**Comment by dinkc64** [comment]
oh - so everything in that image that isn't the game is just a picture?

**Comment by tmaul** [comment]
Yes, specifically this one https://github.com/cosmo0/retropie-overlays-arcade-realistic/blob/master/overlays/configs/batsugun.png The actual cabinet I end up putting a PC in will be much less cool than a real batsugun cabinet, unfortunately :( lol

If you are using full-screen integer scaling on 1080p monitors, it works well to cover up the black borders. Probably would look good on a fancy 4k TV as well.

**Comment by tmaul** [comment]
@dinkc64  @0xe1f @grant2258 @barbudreadmon @iq132  I am thinking of merging this in quite soon (once I have written a readme for sdl2). There are a few things missing (dip switches, input mapping) but I was thinking it would be better to merge it in and then carry on improving it. 

Does that sound reasonable? And does anything look dodgy before I do? I have tried GCC and VS windows builds with this code, as well as SDL1 and 2 builds on linux and windows. I fixed a memory leak in kinst, and had to add a define for M_PI as that seemed to be missing for some reason. As mentioned, I know this has broken MacOS builds as I removed the old interface for cd, but have ported over the latest windows one and it seems to work. Hopefully it doesn't affect anything in libretro builds but that's another reason to consider merging it now

**Comment by barbudreadmon** [comment]
> Hopefully it doesn't affect anything in libretro builds

Don't worry about that, i don't see anything in that PR that could break the libretro port.

**Comment by grant2258** [comment]
Its perfectly usable for me. The paths still need finalized. I was thinking maybe just set the defaults with -createconfig to make a fresh config and use the default variables on this creation instead of setting paths paths in different parts of the code again thats just changes and no reason not to merge it.

**Comment by dinkc64** [comment]
tmaul, it looks good to me :)

**Comment by 0xe1f** [comment]
Sorry I'm a bit late responding here, but yeah, it's cool. I haven't had a chance to check the Pi build, but I'm sure it won't be that big a deal.
Thanks for the heads-up.

**Pull request merged**